libcperciva/datastruct/elasticarray.c

```diff
@@ -25,12 +25,15 @@ resize(struct elasticarray * EA, size_t nsize)
 	void * nbuf;
 
 	/* Figure out how large an allocation we want. */
-	if (EA->alloc < nsize) {
+	if (nsize > EA->alloc) {
 		/* We need to enlarge the buffer. */
-		nalloc = EA->alloc * 2;
```

Comment:
I think the problem here is that the comments don't make it clear what's happening.  (It took me a while to figure it out!)

Let's look at the previous code, imagining that `size_t` can only hold one decimal digit, with the example of:
```
SIZE_MAX = 9;
EA->alloc = 6;
nsize = 8;
```
We then get `nalloc = 6 * 2 = 12 mod 10 = 2`, due to our "simplified example" `size_t` only going up to 9.

The next lines (of the original code) is:
```
if (nalloc < nsize)
        nalloc = nsize;
```

Since we have `2 < 8`, we set `nalloc = 8`.  That's an entirely suitable value.

In other words, the `if` statement has two benefits.  It sets `nalloc` to a good value:
1. if doubling `EA->alloc` is not enough to hold `nsize`
2. if doubling `EA->alloc` "overflows"

I put "overflows" in quotation marks, because according to the C99 specification, it's not technically "overflowing":

> A computation involving unsigned operands can never overﬂow, because a result that cannot be represented by the resulting unsigned integer type is reduced modulo the number that is one greater than the largest value that can be represented by the resulting type.
>
> C99 6.2.5 Types, paragraph 9

If you can think of a counter-example, I'm all ears!  But unless there's a specific example where this code produces a bad value, I think the original code is fine.

It would certainly benefit from a comment, though!  I'm happy to write it, but I'm also happy if you'd like to write one.  I think it needs a comment immediately before the `if (nalloc < nsize)` line, explaining the two cases where it's useful.

Reply:
PS: I forgot to add that `size_t` is definitely unsigned:

> The types are ... size_t which is the unsigned integer type of the result of the sizeof operator
>
> C99 7.17 Common definitions `<stddef.h>`

Reply:
Right. In this change the old code was probably doing the right thing but would have been clearer with a comment. I noticed that too but thought it could be clearer to handle the 'overflow' / 'wrap around' in code more explicitly. 
But if you look at the nsize * 4 below, that one could overflow in a problematic way, right? so I changed both.

Reply:
Hmm.  That's information that would be great to put into the commit message, so that it's clear to everybody that the commit includes both fixes an "overflow" and rewrites other code for clarity.  For tarsnap, we prefer not to mix multiple things in a single commit.  So I would have made one commit for the `nsize * 4` fix, and another for rewriting.

Anyway.  Now that I understand what's happening with the `(EA->alloc < size)`, I don't think the rewriting is necessary, particularly because it adds an extra `if` statement.  I think the best solution is to have a commit which adds a comment as such:

```
/* We need to enlarge the buffer. */
nalloc = EA->alloc * 2;

/*
 * Safely handle if nalloc: 1) is not large enough to hold
 * nsize, or 2) was reduced modulo SIZE_MAX+1.
 */
if (nalloc < nsize)
        nalloc = nsize;
```

Reply:
PS: I'm not 100% sold on the exact wording of that new comment.  In particular, the use of "SIZE_MAX+1" is questionable: I'm not certain that SIZE_MAX must be a power of 2.

(I'm still confident that the right solution is a comment above `if (nalloc < nsize)`, it's just that I'm searching for the best wording of that comment.)

Reply:
Could we instead add more comments above the nalloc = EA->alloc * 2 statement and keep everything else like before. The comment could be something like:
/*
 * We need to enlarge the buffer.
 * Even if the below expression could do modulo wrapping, the nalloc < nsize check underneath should take care of it.
 */
nalloc = EA->alloc * 2;

Reply:
Two quibbles: 1) "should" or "will"?  If we write "should", then that sounds like we're hoping it'll work.  But as Yoda said, "do or do not, there is no ~~try~~ hope".
2) line wrapping; there shouldn't be a newline after "buffer.".  If we really need to have a manual line-break, the comment should start with `/*-`.  (That's a somewhat obscure detail in our `STYLE`.)

Hmm.  What about
> In the case of modulo wrapping, the `if` statement will ensure a sensible default.

I'm not sold on the wording, though.  I still think it's more clear to say:
1. We need to enlarge the buffer.
2. Handle special cases.  (and give details)

but I could be convinced otherwise!

Reply:
Added code comment in [e7500c0](https://github.com/Tarsnap/tarsnap/pull/592/commits/e7500c05ab7c051186d182b210ec694b90b6edc0).